File: andrielfn/day-1/go/santa.go
package main

const (
	up   = 40 // Byte code for "(".
	down = 41 // Byte code for ")".
)

type santa struct {
	floor int
}

// newSanta creates a new Santa presents delivery on floor 0.
func newSanta() *santa {
	return &santa{floor: 0}
}

// move moves the Santa to a different floor based on the given byte.
//
// When byte 40 ("(") is received, Santa goes up.
// When byte 41 (")") is received, Santa goes down.
func (s *santa) move(direction byte) {
	if direction == up {
		s.floor++
	} else if direction == down {
		s.floor--
	}
}

// isOnFloor checks if the Santa is in the given floor.
func (s *santa) isOnFloor(floor int) bool {
	return s.floor == floor
}

File: andrielfn/day-1/elixir/santa.ex
defmodule Santa do
  def move(_, floor \\ 0)

  def move("", floor), do: floor

  def move("(" <> moves, floor) do
    move(moves, floor + 1)
  end

  def move(")" <> moves, floor) do
    move(moves, floor - 1)
  end

  def basement(_, floor \\ 0, count \\ 0)

  def basement(_, -1, count), do: count

  def basement("(" <> moves, floor, count) do
    basement(moves, floor + 1, count + 1)
  end

  def basement(")" <> moves, floor, count) do
    basement(moves, floor - 1, count + 1)
  end
end
